### PR TITLE
feat: adding SafeAccessControlEnumerableCaller interface and implementation

### PR DESCRIPTION
**Pull request by Nilay27**
* Adding interface and implementation for `SafeAccessEnumerableControlCaller()` which will allow any child contract to be able to call methods of `SafeAccessControlEnumerable()`
* Also imported `IAccessControl` from OZ library as `revokeRole()` and `renounceRole()` are not exposed in `SafeAccessControlEnumerable()` but a child contract might need to call these also.

### CODE REVIEW

**Comment by Nilay27** [comment]
* Rebased and updated to include changes from Main.
* Also added natspecs for the interface.

File: packages/prepo-shared-contracts/contracts/SafeAccessControlEnumerableCaller.sol [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    address _safeAccessControlContract,
+    bytes32 _role,
+    address _account
+  ) external view override returns (bool) {
+    return
+      ISafeAccessControlEnumerable(_safeAccessControlContract).isNominated(
+        _role,
+        _account
+      );
+  }
```
Comment (ramenforbreakfast) [inline packages/prepo-shared-contracts/contracts/SafeAccessControlEnumerableCaller.sol:93]: This is missing functions that are added in `SafeAccessControlEnumerable` that are not in `SafeAccessControl`. These should be added so a contract has access to these functions.

File: packages/prepo-shared-contracts/contracts/interfaces/ISafeAccessControlEnumerableCaller.sol [hunk 2]
```diff
@@ -0,0 +1,126 @@
+// SPDX-License-Identifier: AGPL-3.0
+pragma solidity =0.8.7;
+
+/**
+ * @notice This contract enables inheriting contracts to grant, accept,
+ * renounce and revoke access to role or admin role of any SafeAccessControl
```
Comment (ramenforbreakfast) [inline packages/prepo-shared-contracts/contracts/interfaces/ISafeAccessControlEnumerableCaller.sol:6]: There are multiple mentions of `SafeAccessControl` throughout this interface, but it should be `SafeAccessControlEnumerable`

File: packages/prepo-shared-contracts/contracts/interfaces/ISafeAccessControlEnumerableCaller.sol [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+ * contract.
+ */
+interface ISafeAccessControlEnumerableCaller {
+  /**
+   * @notice Nominates a role to be a role admin of a SafeAccessControl
+   * contract.
+   * @dev Only callable by a member of the role's current role admin.
+   * @param safeAccessControlContract Address of the SafeAccessControl
```
Comment (ramenforbreakfast) [inline packages/prepo-shared-contracts/contracts/interfaces/ISafeAccessControlEnumerableCaller.sol:14]: Params must be one line, so we'll have to go for shorter copy throughout the interface.

**Comment by ramenforbreakfast** [comment]
I have addressed my own concerns, the following changes were made:
- Since params can only be one line, shortened the unnecessarily verbose descriptions of the contract parameter for each function.
- Added in the enumerable functions that `SafeAccessControlEnumerable` adds to `SafeAccessControl`.

File: packages/prepo-shared-contracts/contracts/interfaces/ISafeAccessControlEnumerableCaller.sol [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+   * @param safeAccessControlContract Address of the contract to be called
+   * @param role The role of the nomination to be accepted
+   */
+  function acceptRole(address safeAccessControlContract, bytes32 role)
+    external;
+
+  /**
+   * @notice Allows an account to renounce the access to a role for a
+   * SafeAccessControlEnumerable contract.
+   * @dev Only callable by the renouncing address.
```
Comment (XavierEkkel) [inline packages/prepo-shared-contracts/contracts/interfaces/ISafeAccessControlEnumerableCaller.sol:61]:
```suggestion
```

File: packages/prepo-shared-contracts/contracts/interfaces/ISafeAccessControlEnumerableCaller.sol [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+   * contract.
+   * @dev Only callable by the role nominee.
+   * @param safeAccessControlContract Address of the contract to be called
+   * @param role The role of the nomination to be accepted
+   */
+  function acceptRole(address safeAccessControlContract, bytes32 role)
+    external;
+
+  /**
+   * @notice Allows an account to renounce the access to a role for a
```
Comment (XavierEkkel) [inline packages/prepo-shared-contracts/contracts/interfaces/ISafeAccessControlEnumerableCaller.sol:59]:
```suggestion
   * @notice Allows `account` to renounce their current role for a given
```

File: packages/prepo-shared-contracts/contracts/interfaces/ISafeAccessControlEnumerableCaller.sol [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+   * SafeAccessControlEnumerable contract.
+   * @dev Only callable by the renouncing address.
+   * @param safeAccessControlContract Address of the contract to be called
+   * @param role The role for which access is to be renounced
+   * @param account Address of the renouncing account
```
Comment (XavierEkkel) [inline packages/prepo-shared-contracts/contracts/interfaces/ISafeAccessControlEnumerableCaller.sol:64]:
```suggestion
   * @param account Address that is renouncing their role
```

File: packages/prepo-shared-contracts/contracts/interfaces/ISafeAccessControlEnumerableCaller.sol [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    address account
+  ) external;
+
+  /**
+   * @notice Revokes access of an account to a role of a
+   * SafeAccessControlEnumerable contract.
+   * @dev Only callable by a member of the role's current role admin.
+   * @param safeAccessControlContract Address of the contract to be called
+   * @param role The role for which an account's access is to be revoked
+   * @param account Address of the account for which membership is to be revoked
```
Comment (XavierEkkel) [inline packages/prepo-shared-contracts/contracts/interfaces/ISafeAccessControlEnumerableCaller.sol:78]:
```suggestion
   * @param account Address for which membership is to be revoked
```

File: packages/prepo-shared-contracts/contracts/interfaces/ISafeAccessControlEnumerableCaller.sol [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    address account
+  ) external;
+
+  /**
+   * @notice Revokes role nomination for a SafeAccessControlEnumerable
+   * contract.
+   * @dev Only callable by a member of the role's current role admin.
+   * @param safeAccessControlContract Address of the contract to be called
+   * @param role The role of the nomination to be revoked
+   * @param account The address of the nomination to be revoked
```
Comment (XavierEkkel) [inline packages/prepo-shared-contracts/contracts/interfaces/ISafeAccessControlEnumerableCaller.sol:92]:
```suggestion
   * @param account Address for which nomination is to be revoked
```

File: packages/prepo-shared-contracts/contracts/interfaces/ISafeAccessControlEnumerableCaller.sol [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+   */
+  function revokeNomination(
+    address safeAccessControlContract,
+    bytes32 role,
+    address account
+  ) external;
+
+  /**
+   * @param safeAccessControlContract Address of the contract to be called
+   * @param role The role to retrieve a role admin nominee for
```
Comment (XavierEkkel) [inline packages/prepo-shared-contracts/contracts/interfaces/ISafeAccessControlEnumerableCaller.sol:102]:
```suggestion
   * @param role The role to retrieve the role admin nominee for
```

**Review by XavierEkkel — COMMENTED** [review]
the functions need access control modifiers (at least by the inheriting contract), otherwise anyone can call these state-modifying methods

**Comment by Nilay27** [comment]
* Added `onlyOwner` modifiers to all the setters in `AccessControlEnumerableCaller` and `SafeOwnableCaller` to ensure that only the owner of the contract/inheriting contract can call those setters.
* Made changes to `SafeOwnableCaller` in this PR only as it was already merged and the changes were small.

**Comment by Nilay27** [comment]
* Made `SafeOwnableCaller` and `SafeAccessControlEnumerablCaller` abstract.
* Included changes for `SafeOwnableCaller` in this PR only as they were small and only required changing visibility from `external` to `public' and making methods `virtual`
* Included dev comment in the interface for both that inheriting contracts must override the methods with appropriate modifiers.

**Review by ramenforbreakfast — APPROVED** [review]
I rewrote the dev comment clarifying that access modifiers must be added to be clearer and make `ISafeOwnableCaller` and `ISafeAccessControlEnumerableCaller` consistent with each other.

**Pull request merged**